### ISSUE TITLE
Use of aggregates without groupping

### ISSUE THREAD

**Issue by MaceWindu** [linq2db/linq2db#679]
I've tried to write a query for this question (edit: there was link before, but StackOverflow removed question) and found several issues. I'll fill separate defects for each.

BTW, maybe we should create gitter channel for development-related discussions?

Test code for all issues:

```cs
[Table("Table", Database = "Database", Schema = "Schema")]
class TestTable
{
	[Column]
	public DateTime SentTimestamp;
}

[DataContextSource]
public void Test(string context)
{
	using (var db = GetDataContext(context))
	{
		var now = DateTime.Now;
		var weekStart = now.Date.AddDays(DayOfWeek.Monday - now.DayOfWeek);
		var monthStart = new DateTime(now.Year, now.Month, 1);
		var yearStart = new DateTime(now.Year, 1, 1);

		var result = (from record in db.GetTable<TestTable>()
			// for current issue
			group record by 1 into g
			// for related issues
			// group record by record.SentTimestamp into g
			select new
			{
				ThisWeek = g.Count(r => r.SentTimestamp > weekStart),
				ThisMonth = g.Count(r => r.SentTimestamp > monthStart),
				Last90Days = g.Count(r => r.SentTimestamp > now.AddDays(-90)),
				ThisYear = g.Count(r => r.SentTimestamp > yearStart),
				Last6Months = g.Count(r => r.SentTimestamp > now.AddMonths(-6)),
			}).FirstOrDefault();
	}
}
```

linq2db fails to generate SQL for code above:
```cs
LinqToDB.LinqToDBException : 'r.SentTimestamp' cannot be converted to SQL.
```

**Comment by sdanyliv** [comment]
@MaceWindu, try to use Window Functions without grouping, not checked yet but i think it should work.
`Sql.Ext.Count(expression).ToValue()`

**Comment by MaceWindu** [comment]
Great!
It works:

```
var result = (from record in db.GetTable<TestTable>()
	  select new
	  {
		  ThisWeek = Sql.Ext.Count(record.SentTimestamp > weekStart ? 1 : (int?)null).ToValue(),
		  ThisMonth = Sql.Ext.Count(record.SentTimestamp > monthStart ? 1 : (int?)null).ToValue(),
		  Last90Days = Sql.Ext.Count(record.SentTimestamp > now.AddDays(-90) ? 1 : (int?)null).ToValue(),
		  ThisYear = Sql.Ext.Count(record.SentTimestamp > yearStart ? 1 : (int?)null).ToValue(),
		  Last6Months = Sql.Ext.Count(record.SentTimestamp > now.AddMonths(-6) ? 1 : (int?)null).ToValue()
	  }).FirstOrDefault();
```

```
DECLARE @weekStart DateTime2
SET     @weekStart = '2017-05-22'
DECLARE @monthStart DateTime2
SET     @monthStart = '2017-05-01'
DECLARE @p1 DateTime2
SET     @p1 = '2017-02-23T21:42:48'
DECLARE @yearStart DateTime2
SET     @yearStart = '2017-01-01'
DECLARE @p2 DateTime2
SET     @p2 = '2016-11-24T21:42:48'

SELECT TOP (1)
	COUNT(CASE
		WHEN [t1].[SentTimestamp] > @weekStart
			THEN 1
		ELSE NULL
	END) as [c1],
	COUNT(CASE
		WHEN [t1].[SentTimestamp] > @monthStart
			THEN 1
		ELSE NULL
	END) as [c2],
	COUNT(CASE
		WHEN [t1].[SentTimestamp] > @p1
			THEN 1
		ELSE NULL
	END) as [c3],
	COUNT(CASE
		WHEN [t1].[SentTimestamp] > @yearStart
			THEN 1
		ELSE NULL
	END) as [c4],
	COUNT(CASE
		WHEN [t1].[SentTimestamp] > @p2
			THEN 1
		ELSE NULL
	END) as [c5]
FROM
	[Table] [t1]
```

**Comment by sdanyliv** [comment]
@MaceWindu, i will extend (i had plans) current functions tho work with gouping, but i need some digging in current Builders.

**Comment by sdanyliv** [comment]
Fixed in #775

For now you can use the following syntax
```c#
		var result = (from record in db.GetTable<TestTable>()
			// for current issue
			group record by 1 into g
			// for related issues
			// group record by record.SentTimestamp into g
			select new
			{
				ThisWeek = g.CountExt(r => r.SentTimestamp > weekStart ? 1 : (int?)null),
				ThisMonth = g.CountExt(r => r.SentTimestamp > monthStart ? 1 : (int?)null),
				Last90Days = g.CountExt(r => r.SentTimestamp > now.AddDays(-90) ? 1 : (int?)null),
				ThisYear = g.CountExt(r => r.SentTimestamp > yearStart ? 1 : (int?)null),
				Last6Months = g.CountExt(r => r.SentTimestamp > now.AddMonths(-6) ? 1 : (int?)null),
			}).FirstOrDefault();
```
